odxtools/diaglayer.py

```diff
@@ -256,60 +338,141 @@ def diag_data_dictionary_spec(self) -> DiagDataDictionarySpec:
-            for dop in parent_ref.get_inherited_data_object_properties():
-                result_dict[dop.short_name] = dop
+    def _get_local_diag_comms(self, odxlinks: OdxLinkDatabase
+                             ) -> Iterable[Union[DiagService, SingleEcuJob]]:
+        """Return the list of locally defined diagnostic communications.
 
-        if self.local_diag_data_dictionary_spec:
-            for dop in self.local_diag_data_dictionary_spec.all_data_object_properties:
-                result_dict[dop.short_name] = dop
+        This is not completely trivial as it requires to resolving the
+        references specified in the <DIAG-COMMS> XML tag.
+        """
+        result_dict: Dict[str, Union[DiagService, SingleEcuJob]] = {}
+
+        # TODO (?): add objects from the import-refs
 
-        return list(result_dict.values())
+        for dc_proxy in self.diag_layer_raw.diag_comms:
+            if isinstance(dc_proxy, OdxLinkRef):
+                dc = odxlinks.resolve(dc_proxy)
+            else:
+                dc = dc_proxy
+
+            assert isinstance(dc, (DiagService, SingleEcuJob))
+            assert dc.short_name not in result_dict, (
+                f"Multiple definitions of DIAG-COMM '{dc.short_name}' in "
+                f"layer '{self.short_name}'")
+            result_dict[dc.short_name] = dc
+
+        return result_dict.values()
+
+    def _get_local_unit_groups(self) -> Iterable[UnitGroup]:
+        if self.diag_layer_raw.diag_data_dictionary_spec is None:
+            return []
+
+        unit_spec = self.diag_layer_raw.diag_data_dictionary_spec.unit_spec
+        if unit_spec is None:
+            return []
+
+        return unit_spec.unit_groups
+
+    def _compute_available_diag_comms(self, odxlinks: OdxLinkDatabase
+                                     ) -> Iterable[Union[DiagService, SingleEcuJob]]:
+        get_local_objects_fn = lambda dl: dl._get_local_diag_comms(odxlinks)
+        not_inherited_fn = lambda parent_ref: parent_ref.not_inherited_diag_comms
+        return self._compute_available_objects(get_local_objects_fn, not_inherited_fn)
+
+    def _compute_available_global_neg_responses(self, odxlinks: OdxLinkDatabase) \
+            -> Iterable[Response]:
+        get_local_objects_fn = lambda dl: dl.diag_layer_raw.global_negative_responses
+        not_inherited_fn = lambda parent_ref: parent_ref.not_inherited_global_neg_responses
+        return self._compute_available_objects(get_local_objects_fn, not_inherited_fn)
+
+    def _compute_available_data_object_props(self) -> Iterable[DataObjectProperty]:
+        get_local_objects_fn = lambda dl: \
+            [] if dl.diag_layer_raw.diag_data_dictionary_spec is None \
+            else dl.diag_layer_raw.diag_data_dictionary_spec.data_object_props
+        not_inherited_fn = lambda parent_ref: parent_ref.not_inherited_dops
+        return self._compute_available_objects(get_local_objects_fn, not_inherited_fn)
```

Comment:
Why the lambda? why not just pass the parent_ref.not_inherited_dops list?

Reply:
because it is possible to have nested inheritance hierarchies where each diaglayer specifies multiple parentrefs...